tests/test_injection.py

```diff
@@ -535,6 +542,126 @@ def callback(
         # not injected at all.
         assert set(kwargs.keys()) == {"expensive", "item"}
 
+    @inlineCallbacks
+    def test_dynamic_deps(self):
+        def callback(dd: DynamicDeps):
+            pass
+
+        provider = get_provider({Cls1, Cls2})
+        injector = get_injector_for_testing({provider: 1})
+
+        expected_instances = {
+            DynamicDeps: DynamicDeps({Cls1: Cls1(), Cls2: Cls2()}),
+            Cls1: Cls1(),
+            Cls2: Cls2(),
+        }
+        expected_kwargs = {
+            "dd": DynamicDeps({Cls1: Cls1(), Cls2: Cls2()}),
+        }
+        yield self._assert_instances(
+            injector,
+            callback,
+            expected_instances,
+            expected_kwargs,
+            reqmeta={"inject": [Cls1, Cls2]},
+        )
+
+    @inlineCallbacks
+    def test_dynamic_deps_mix(self):
+        def callback(c1: Cls1, dd: DynamicDeps):
+            pass
+
+        provider = get_provider({Cls1, Cls2})
+        injector = get_injector_for_testing({provider: 1})
+
+        expected_instances = {
+            DynamicDeps: DynamicDeps({Cls1: Cls1(), Cls2: Cls2()}),
+            Cls1: Cls1(),
+            Cls2: Cls2(),
+        }
```

Comment:
In this test DynamicDeps[Cls1] is not the same instance as Cls1 instance received in the callback.

Is it an artifact of the test, or is it the way the implementation works?

Does it cause any problems, and if so, should they be fixed or documented?

I guess there could be 2 kinds of issues. First, you may just get the same data, but in different instances (i.e. copies). Second, you may need to do some work multiple times (fetch web page twice?? Run page object extraction code twice to get the same item?). The second issue looks unlikely (right?), but the first seems to apply - should we document it, at least briefly?

Reply:
It's actually the same instance, these checks in several tests (including old ones) aren't really correct as Cls1 and Cls2 are strings and so are compared just by the value. If this test was checking instance equality the expectation would need to be rewritten.

Reply:
Nice!
Does it worth a test, or is it already tested somewhere?